File: src/application/useCases/sharedMerchants.ts
import type { Id, Merchant, Transaction } from '../../domain/entities/types'
import {
  baselineCatalog,
  contributionFor,
  effectiveEntry,
  latestUpdate,
  planAccountMerchantSync,
  shareableName,
  sharedMerchantKey,
  type SharedMerchantEntry,
} from '../../domain/sharedMerchantCatalog'
import type { Clock, MerchantRepository } from '../ports/repositories'
import type { SharedMerchantCatalogPort, SyncCursorPort } from '../ports/SharedMerchantCatalogPort'

export interface SharedMerchantsDeps {
  catalog: SharedMerchantCatalogPort
  merchants: MerchantRepository
  /** تجار المرجع اللي جوه التطبيق — أساس القاعدة (OVERRIDES §25.1). */
  baseline: readonly Merchant[]
  /** معرّفات شجرة التصنيفات بس — نفس المعرّف في كل الحسابات. */
  treeCategoryIds: ReadonlySet<Id>
  cursor: SyncCursorPort
  /** آخر مراجعة لكل المؤكدين — التأكيد من لوحة فايربيز ممكن ما يغيّرش `updatedAt` فما يوصلش بالتغييرات لوحدها. */
  confirmedCursor: SyncCursorPort
  clock: Clock
}

/** كل المؤكدين بيتقروا مرة في اليوم على الأكتر (حد القراية المجانية 50 ألف في اليوم). */
const CONFIRMED_REFRESH_MS = 24 * 60 * 60 * 1000

/**
 * قاعدة التجار المشتركة — OVERRIDES §25 و§25.1.
 * - `sync`: بيقرا التغييرات من آخر مرة، **وكل المؤكدين مرة في اليوم** (عشان تأكيد اتعمل من اللوحة من غير ما
 *   `updatedAt` يتغير يوصل برضه)، وبيضيف للحساب المؤكد الناقص من غير ما يكتب فوق تصنيف المستخدم.
 * - `contribute`: لما المستخدم يأكد تصنيف عملية شراء، اسم المحل وتصنيفه بيتبعتوا مش مؤكدين.
 */
export function makeSharedMerchants(deps: SharedMerchantsDeps) {
  const baseline = baselineCatalog(deps.baseline)

  async function sync(): Promise<{ changes: number; added: number; filled: number }> {
    const since = deps.cursor.read()
    const now = deps.clock.nowIso()
    const last = deps.confirmedCursor.read()
    // أول مزامنة بتقرا كله أصلًا؛ بعدها كل المؤكدين مرة في اليوم
    const refresh = since !== null && (!last || Date.parse(now) - Date.parse(last) >= CONFIRMED_REFRESH_MS)
    const [changed, confirmed] = await Promise.all([
      deps.catalog.listChangedSince(since),
      refresh ? deps.catalog.listConfirmed() : Promise.resolve([] as SharedMerchantEntry[]),
    ])
    const byKey = new Map(confirmed.map((e) => [sharedMerchantKey(e.normalizedName), e]))
    for (const e of changed) byKey.set(sharedMerchantKey(e.normalizedName), e)
    const remote = [...byKey.values()]
    let added = 0
    let filled = 0
    if (remote.length > 0) {
      const effective = remote.map((r) => effectiveEntry(baseline.get(sharedMerchantKey(r.normalizedName)), r)!)
      const plan = planAccountMerchantSync(effective, await deps.merchants.listAll(), deps.treeCategoryIds)
      const writes = [...plan.add, ...plan.fill]
      if (writes.length > 0) await deps.merchants.saveMany(writes)
      added = plan.add.length
      filled = plan.fill.length
    }
    // المؤشرات بتتقدم بعد الكتابة بس — انقطاع في النص بيعيد نفس التغييرات (المعرّفات ثابتة فمفيش تكرار)
    const latest = latestUpdate(changed, since)
    if (latest && latest !== since) deps.cursor.write(latest)
    if (since === null || refresh) deps.confirmedCursor.write(now)
    return { changes: remote.length, added, filled }
  }

  async function contribute(
    transaction: Pick<Transaction, 'economicKind' | 'observedDirection' | 'rawMerchantName'>,
    categoryId: Id,
  ): Promise<'shared' | 'skipped'> {
    // الفحص الرخيص الأول: مفيش قراءة من السيرفر لعملية مش هتتبعت أصلًا
    const probe = contributionFor({ transaction, categoryId, treeCategoryIds: deps.treeCategoryIds, current: undefined })
    const normalized = shareableName(transaction.rawMerchantName)
    if (!probe || !normalized) return 'skipped'
    const current = effectiveEntry(baseline.get(sharedMerchantKey(normalized)), await deps.catalog.get(normalized))
    const entry = contributionFor({ transaction, categoryId, treeCategoryIds: deps.treeCategoryIds, current })
    if (!entry) return 'skipped'
    await deps.catalog.save(entry)
    return 'shared'
  }

  return { sync, contribute }
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/SharedMerchants.kt
package app.masroufy.core

/**
 * قاعدة التجار المشتركة — نقل `src/domain/sharedMerchantCatalog.ts` (OVERRIDES §25 و§25.1).
 * بيتشارك الاسم وأشكاله وتصنيف من الشجرة وحالة التأكيد بس — مفيش مبالغ ولا تواريخ ولا مين ضاف.
 */
data class SharedMerchantEntry(
    val normalizedName: String,
    val displayName: String,
    val aliases: List<String>,
    /** تصنيف من شجرة التطبيق بس. */
    val categoryId: Id?,
    /** بيتكتب من لوحة فايربيز بس — التطبيق عمره ما يكتبه true. */
    val confirmed: Boolean,
    val updatedAt: String? = null,
)

const val MAX_SHARED_NAME = 80
const val MAX_SHARED_ALIASES = 10

/** مفتاح المستند: الاسم الموحد بشرطة مكان المسافة. */
fun sharedMerchantKey(normalizedName: String): String = normalizeText(normalizedName).replace(' ', '-')

/** 4 أرقام ورا بعض (أي نوع أرقام) = شكل حساب أو تليفون ⇒ ما يتبعتش (قاعدة 11). */
private val LONG_DIGITS = Regex("\\p{N}{4,}")

/** ينفع يتشارك؟ فاضي أو طويل أو فيه رقم طويل ⇒ null. */
fun shareableName(name: String?): String? {
    val normalized = normalizeText(name ?: "")
    if (normalized.isEmpty() || normalized.length > MAX_SHARED_NAME || LONG_DIGITS.containsMatchIn(normalized)) return null
    return normalized
}

/** المرجع اللي جوه التطبيق كأساس: التاجر المؤكد فيه بيبدأ مؤكد. */
fun baselineCatalog(merchants: List<Merchant>): Map<String, SharedMerchantEntry> {
    val out = LinkedHashMap<String, SharedMerchantEntry>()
    for (m in merchants) {
        out[sharedMerchantKey(m.normalizedName)] = SharedMerchantEntry(
            m.normalizedName, m.displayName, m.aliases.orEmpty(), m.verifiedCategoryId, !m.verifiedCategoryId.isNullOrEmpty(),
        )
    }
    return out
}

/** تعديل مش مؤكد ما يغلبش تاجر مؤكد في المرجع. */
fun effectiveEntry(baseline: SharedMerchantEntry?, remote: SharedMerchantEntry?): SharedMerchantEntry? {
    if (remote == null) return baseline
    if (baseline?.confirmed == true && !remote.confirmed) return baseline
    return remote
}

/** مساهمة المستخدم لما يأكد تصنيف **شراء**؛ null = مفيش حاجة تتبعت (والمؤكد ما يتغيرش إلا من اللوحة). */
fun contributionFor(
    economicKind: EconomicKind,
    observedDirection: Direction,
    rawMerchantName: String?,
    categoryId: Id,
    treeCategoryIds: Set<Id>,
    current: SharedMerchantEntry?,
): SharedMerchantEntry? {
    if (economicKind != EconomicKind.PURCHASE || observedDirection != Direction.OUT) return null
    if (categoryId !in treeCategoryIds) return null
    val normalized = shareableName(rawMerchantName) ?: return null
    if (current?.confirmed == true) return null
    if (current != null && current.categoryId == categoryId) return null
    val displayName = JsText.collapseWhitespace(JsText.trim(rawMerchantName ?: "")).take(MAX_SHARED_NAME)
    return SharedMerchantEntry(normalized, current?.displayName ?: displayName, (current?.aliases ?: emptyList()).take(MAX_SHARED_ALIASES), categoryId, false)
}

data class AccountMerchantSyncPlan(val add: List<Merchant>, val fill: List<Merchant>)

/** **المؤكد بس** بيوصل الحساب، وتصنيف المستخدم نفسه عمره ما يتكتب فوقه؛ المعرّف ثابت من الاسم. */
fun planAccountMerchantSync(entries: List<SharedMerchantEntry>, account: List<Merchant>, treeCategoryIds: Set<Id>): AccountMerchantSyncPlan {
    val index = merchantIndex(account)
    val add = mutableListOf<Merchant>()
    val fill = mutableListOf<Merchant>()
    val seen = mutableSetOf<String>()
    for (entry in entries) {
        val normalized = normalizeText(entry.normalizedName)
        if (!entry.confirmed || entry.categoryId.isNullOrEmpty() || entry.categoryId !in treeCategoryIds) continue
        if (normalized.isEmpty() || !seen.add(normalized)) continue
        val existing = index[normalized]
        if (existing != null) {
            if (existing.verifiedCategoryId.isNullOrEmpty()) fill += existing.copy(verifiedCategoryId = entry.categoryId)
            continue
        }
        add += Merchant(
            id = "merch-shared-${sharedMerchantKey(normalized)}",
            displayName = entry.displayName.ifEmpty { normalized },
            normalizedName = normalized,
            aliases = entry.aliases.takeIf { it.isNotEmpty() }?.take(MAX_SHARED_ALIASES),
            verifiedCategoryId = entry.categoryId,
        )
    }
    return AccountMerchantSyncPlan(add, fill)
}

/** أحدث وقت تعديل — منه المزامنة الجاية بتبدأ. */
fun latestUpdate(entries: List<SharedMerchantEntry>, previous: String?): String? {
    var latest = previous
    for (e in entries) if (!e.updatedAt.isNullOrEmpty() && (latest.isNullOrEmpty() || e.updatedAt > latest)) latest = e.updatedAt
    return latest
}
